File: lua/nucleotide-count/nucleotide-count.lua
local DNA = {}
DNA.__index = DNA

function DNA:new(strand)
    local counts = { A=0, T=0, C=0, G=0 }
    for char in (strand or ""):gmatch(".") do
        assert(counts[char], "Invalid Sequence")
        counts[char] = counts[char] + 1
    end

    local dna = { nucleotideCounts = counts }
    setmetatable(dna, self)
    return dna
end

function DNA:count(nucleotide)
    local n = self.nucleotideCounts[nucleotide]
    assert(n, "Invalid Nucleotide")
    return n
end

return DNA

File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def init-nucleotide-count {\A 0 \C 0 \G 0 \T 0})

(defn validate-nucleotide [nucleotide]
  (when-not (find init-nucleotide-count nucleotide)
    (throw (IllegalArgumentException. "invalid nucleotide"))))

(defn validate-strand [strand]
  (loop [[h & t] strand]
    (when h
      (validate-nucleotide h)
      (recur t))))

(defn nucleotide-counts [strand]
  (validate-strand strand)
  (->> (frequencies strand)
       (merge init-nucleotide-count)))

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (validate-nucleotide nucleotide)
  (-> (nucleotide-counts strand)
      (get nucleotide)))
